contracts/yieldaggregator-bonus/src/execute/distribution.rs

```diff
@@ -0,0 +1,273 @@
+) -> StdResult<HashMap<Addr, Coins>> {
+    let mut rewards = HashMap::<Addr, Coins>::new();
+    
+    for vault_id in vault_ids {
+        let price = price_data
+            .iter()
+            .find(|item| item.0 == vault_id)
+            .unwrap()
+            .1;
+
+        let mut rewards_in_vault_id = VAULT_SHARE_STAKINGS
+            .prefix((bonus_window_id, vault_id.to_owned()))
+            .range(store, None, None, Order::Ascending)
+            .map(|item| -> StdResult<_> {
+                let (addr, staking) = item?;
+
+                let proportion_of_staking = Decimal::from_ratio(staking.vault_share, Uint128::one())
+                    .checked_mul(*price)?
+                    .checked_div(normalized_total_staked_amount_in_same_unit)
+                    .unwrap();
+                let mut total_reward = Decimal::to_uint_floor(Decimal::from_ratio(total_potential_reward, Uint128::one())
+                    .checked_mul(proportion_of_staking)?);
+
+                if voting_winners.keys().to_owned().collect::<Vec<&VaultId>>().contains(&&vault_id) {
+                    let all_reward_in_votintg= voting_winners.get(&vault_id).unwrap();
+                    let proportional_reward_in_voting = staking.staking_power_index
+                        .checked_div(total_staking_info_for_all.get(&vault_id).unwrap().1)
+                        .unwrap()
+                        .checked_mul(Decimal::from_ratio(*all_reward_in_votintg, Uint128::one()))?;
+                        
+                    total_reward = total_reward.checked_add(Decimal::to_uint_floor(proportional_reward_in_voting))?;
+                }
+
+                let mut coins = Coins::default();
+                let vault_share_denom = format!("yieldaggregator/vault/{}", vault_id);
+                coins.add(Coin::new(total_reward.into(), "".to_string()))?;
+                coins.add(Coin::new(staking.vault_share.into(), vault_share_denom))?;
+        
+                Ok((addr, coins))
+            })
+            .collect::<StdResult<HashMap<Addr, Coins>>>()?;
+
+        rewards.extend(rewards_in_vault_id.drain());
+    }
+
+    Ok(rewards)
+}
+
+// Create SendMsg for each reward receiver
+pub fn create_send_msg_for_rewards(
+    rewards: &HashMap<Addr, Coins>,
+    reward_token_denom: String,
+) -> StdResult<Vec<CosmosMsg>> {
+    let mut send_msgs = Vec::<CosmosMsg>::new();
+    for (addr, coins) in rewards {
+        let new_coins = coins.clone().into_vec().iter().map(|coin| {
+            if coin.denom == "" {
+                let new_coin = Coin::new(coin.amount.into(), reward_token_denom.clone());
+                return new_coin
+            }
+            coin.to_owned()
+        }).collect::<Vec<Coin>>();
+
+        let send_msg = CosmosMsg::Bank(BankMsg::Send {
+            to_address: addr.to_string(),
+            amount: new_coins,
+        });
+        send_msgs.push(send_msg);
+    }
+    Ok(send_msgs)
```

Comment:
Is it fine to distribute the reward on a single message? I think it would overflow the block gas limit if the number of users are more than 1000